src/chan_vese.jl

```diff
@@ -0,0 +1,254 @@
+for each step, serves to accelerate the algorithm. Although larger `Δt`
+can speed up the algorithm, it might prevent algorithm from converging to 
+the solution.
+
+Default: 0.5
+
+## normalize::Bool
+
+The arguement `normalize` controls whether to normalize the input img.
+
+Default: false
+
+## init_level_set
+
+The arguement `init_level_set` allows users to initalize the level set 𝚽⁰, 
+whose size should be the same as input image.
+
+If the default init level set is uesd, it will be defined as:
+𝚽⁰ₓ = ∏ sin(xᵢ ⋅ π / 5), where xᵢ are pixel coordinates, i = 1, 2, ⋯, ndims(img). 
+This level set has fast convergence, but may fail to detect implicit edges.
+
+Default: initial_level_set(size(img))
+
+# Examples
+
+```julia
+using TestImages
+using ImageSegmentation
+
+img = testimage("cameraman")
+
+cv_result = chan_vese(img, max_iter=200)
+```
+
+# References
+
+[1] An Active Contour Model without Edges, Tony Chan and Luminita Vese, 
+    Scale-Space Theories in Computer Vision, 1999, :DOI:`10.1007/3-540-48236-9_13`
+[2] Chan-Vese Segmentation, Pascal Getreuer Image Processing On Line, 2 (2012), 
+    pp. 214-224, :DOI:`10.5201/ipol.2012.g-cv`
+[3] The Chan-Vese Algorithm - Project Report, Rami Cohen, 2011 :arXiv:`1107.2782`
+"""
+function chan_vese(img::GenericGrayImage;
+                   μ::Real=0.25,
+                   λ₁::Real=1.0,
+                   λ₂::Real=1.0,
+                   tol::Real=1e-3,
+                   max_iter::Int=500,
+                   Δt::Real=0.5,
+                   normalize::Bool=false,
+                   init_level_set=initial_level_set(size(img)))
+    # Signs used in the codes and comments mainly follow paper[3] in the References.
+    axes(img) == axes(init_level_set) || throw(ArgumentError("axes of input image and init_level_set should be equal. Instead they are $(axes(img)) and $(axes(init_level_set))."))
+    img = float64.(channelview(img))
+    N = ndims(img)
+    iter = 0
+    h = 1.0
+    del = tol + 1
+    if normalize
+        img .= img .- minimum(img)
+        if maximum(img) != 0
+            img .= img ./ maximum(img)
+        end
+    end
+
+    # Precalculation of some constants which helps simplify some integration
+    area = length(img)   # area = ∫H𝚽 + ∫H𝚽ⁱ
+    ∫u₀ = sum(img)       # ∫u₀ = ∫u₀H𝚽 + ∫u₀H𝚽ⁱ
+
+    # Initialize the level set
+    𝚽ⁿ = init_level_set
+
+    # Preallocation and initializtion
+    H𝚽 = trues(size(img)...)
+    𝚽ⁿ⁺¹ = similar(𝚽ⁿ)
+
+    Δ = ntuple(d -> CartesianIndex(ntuple(i -> i == d ? 1 : 0, N)), N)
+    idx_first = first(CartesianIndices(𝚽ⁿ))
+    idx_last = last(CartesianIndices(𝚽ⁿ))
+    
+    while (del > tol) & (iter < max_iter)
+        ϵ = 1e-8
+        diff = 0
+
+        # Calculate the average intensities
+        @. H𝚽 = 𝚽ⁿ > 0 # Heaviside function
+        c₁, c₂ = calculate_averages(img, H𝚽, area, ∫u₀) # Compute c₁(𝚽ⁿ), c₂(𝚽ⁿ)
+
+        # Calculate the variation of level set 𝚽ⁿ
+        @inbounds @simd for idx in CartesianIndices(𝚽ⁿ)
```

Comment:
`@simd` won't work if the function is complicated (e.g., has if-branch and uninlined functions)

```suggestion
        @inbounds for idx in CartesianIndices(𝚽ⁿ)
```